### ISSUE TITLE
带权重的tversky 替换 交叉熵 loss，预测结果为黑图， 求看？

### ISSUE THREAD

**Issue by pauperonway** [MaybeShewill-CV/lanenet-lane-detection#403]
**_修改文件为：lanenet_back_end.py
定义处：_**

    def tversky(self, prediction, ground_truth, weight_map=None, alpha=0.7, beta=0.3):
        prediction = tf.to_float(prediction)
        if len(ground_truth.shape) == len(prediction.shape):
            ground_truth = ground_truth[..., -1]
        one_hot = self.labels_to_one_hot(ground_truth, tf.shape(prediction)[-1])
        one_hot = tf.sparse_tensor_to_dense(one_hot)
        p0 = prediction
        p1 = 1 - prediction
        g0 = one_hot
        g1 = 1 - one_hot
        if weight_map is not None:
            weight_map_nclasses = tf.multiply(one_hot, weight_map)
        else:
            weight_map_nclasses = 1

        tp = tf.reduce_sum(weight_map_nclasses * p0 * g0)
        fp = alpha * tf.reduce_sum(weight_map_nclasses * p0 * g1)
        fn = beta * tf.reduce_sum(weight_map_nclasses * p1 * g0)
        EPSILON = 0.00001
        numerator = tp
        denominator = tp + fp + fn + EPSILON
        score = numerator / denominator
        return 1.0 - tf.reduce_mean(score)

**_调用处：_**

               '''
                binary_segmenatation_loss = self._compute_class_weighted_cross_entropy_loss(
                    onehot_labels=binary_label_onehot,
                    logits=binary_seg_logits,
                    classes_weights=inverse_weights
                )
                '''
                binary_segmenatation_loss = self.tversky(prediction = binary_seg_logits, ground_truth = binary_label, weight_map=inverse_weights, alpha=0.5, beta=0.5)

**Comment by pauperonway** [comment]
tversky loss 参考：https://github.com/NifTK/NiftyNet/blob/dev/niftynet/layer/loss_segmentation.py

**Comment by MaybeShewill-CV** [comment]
@pauperonway tversky loss没做过实验。dice我实验过，还是可以提升精度的:)

**Comment by pauperonway** [comment]
**_@MaybeShewill-CV 我的dice函数也是黑图, 好像是权重的问题，但是还是看不出问题···_**

def dice(self, prediction, onehot, weight_map=None):
        # Dice as according to the paper:
        prediction = tf.cast(prediction, tf.float32)
        num_classes = tf.shape(prediction)[-1]
        one_hot = onehot    #self.labels_to_one_hot(ground_truth, num_classes=num_classes)
        softmax_of_logits = tf.nn.softmax(prediction)
        if weight_map is not None:
            #weight_map_nclasses = tf.tile(tf.reshape(weight_map, [-1, 1]), [1, num_classes])
            weight_map_nclasses = weight_map
            dice_numerator = 2.0 * tf.reduce_sum(
                weight_map_nclasses * one_hot * softmax_of_logits)
            dice_denominator = \
                tf.reduce_sum(weight_map_nclasses * softmax_of_logits) + \
                tf.reduce_sum(weight_map_nclasses * one_hot)
        else:
            dice_numerator = 2.0 * tf.reduce_sum(
                one_hot * softmax_of_logits)
            dice_denominator = \
                tf.reduce_sum(softmax_of_logits) + \
                tf.reduce_sum(one_hot)

        epsilon = 0.00001
        loss_dice = -(dice_numerator + epsilon) / (dice_denominator + epsilon)

        return loss_dice

**Comment by MaybeShewill-CV** [comment]
@pauperonway 你可以把这个weight_map去掉:)

**Comment by pauperonway** [comment]
@MaybeShewill-CV 我尝试使用不加权重的交叉熵，预测出全黑图，使用不加权重的dice也是全黑图。感谢这么快回复！～

**Comment by MaybeShewill-CV** [comment]
@pauperonway 不加权重的交叉熵也不会是全黑图，你查看结果如果是自己的数据的话不要使用后处理 只检查binary segmentation的结果就好。dice的话可能需要你检查自己的实现有没有问题。我使用bisenet做back end时候dice没有问题，这部分你可以参考https://github.com/MaybeShewill-CV/bisenetv2-tensorflow/blob/1a9ced8d257d113272a5f15d11227003719e195c/bisenet_model/bisenet.py#L887-L955

**Comment by pauperonway** [comment]
我一般训练500epoch后，加权重的entropy loss预测一张有车道线的图统计前景点有值，不加权重就全0。dice加不加权重都是0。

**Comment by pauperonway** [comment]
我使用了自己标注的数据集，是间断的实际车道线，不平衡较大。所以使用dice或者没有加权重的entropy loss会全部预测为背景。（自己的一种猜测，因为加权重就可以正常。）放弃使用dice。我用deeplabv3+训练的车道线效果很好，但是deeplab不能实现实例分割，后面会持续改进lanenet···

**Comment by MaybeShewill-CV** [comment]
@pauperonway dice应对的就是这种不平衡的问题. 其实deeplab v3也可以改造成lanenet的这种结构，我之前实验过效果还可以但是deeplabv3太慢了 最后就放弃这个backbone了。其实可以试试bisenet或者bisenetv2 个人很喜好这两个工作 哈哈:)

**Comment by pauperonway** [comment]
@MaybeShewill-CV 我现在使用修改后的语义分割进行多类别分割，实例网络进行左右车道线聚类（算多任务网络:) ）。使用bisenetv2可以增加一个实例分割分支吗？让我实现多任务的end-to-end.....

**Comment by MaybeShewill-CV** [comment]
@pauperonway 可以增加一个instance seg的分支啊。
